webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/ZarrAgglomerateService.scala

```diff
@@ -0,0 +1,291 @@
+import com.scalableminds.webknossos.datastore.geometry.Vec3IntProto
+import com.scalableminds.webknossos.datastore.helpers.{NativeBucketScanner, NodeDefaults, SkeletonTracingDefaults}
+import com.scalableminds.webknossos.datastore.models.datasource.{DataSourceId, ElementClass}
+import com.scalableminds.webknossos.datastore.storage.{AgglomerateFileKey, DataVaultService, RemoteSourceDescriptor}
+import com.typesafe.scalalogging.LazyLogging
+import net.liftweb.common.Box.tryo
+import ucar.ma2.{Array => MultiArray}
+
+import java.nio.{ByteBuffer, ByteOrder, LongBuffer}
+import scala.collection.compat.immutable.ArraySeq
+import scala.concurrent.ExecutionContext
+
+class ZarrAgglomerateService(config: DataStoreConfig,
+                             dataVaultService: DataVaultService,
+                             sharedChunkContentsCache: AlfuCache[String, MultiArray])
+    extends DataConverter
+    with LazyLogging {
+
+  private lazy val openArraysCache = AlfuCache[(AgglomerateFileKey, String), DatasetArray]()
+
+  def clearCache(predicate: ((AgglomerateFileKey, String)) => Boolean): Int =
+    openArraysCache.clear(predicate)
+
+  protected lazy val bucketScanner = new NativeBucketScanner()
+
+  private val keySegmentToAgglomerate = "segment_to_agglomerate"
+  private val keyAgglomerateToSegmentsOffsets = "agglomerate_to_segments_offsets"
+  private val keyAgglomerateToSegments = "agglomerate_to_segments"
+  private val keyAgglomerateToPositions = "agglomerate_to_positions"
+  private val keyAgglomerateToEdges = "agglomerate_to_edges"
+  private val keyAgglomerateToEdgesOffsets = "agglomerate_to_edges_offsets"
+  private val keyAgglomerateToAffinities = "agglomerate_to_affinities"
+
+  private def mapSingleSegment(segmentToAgglomerate: DatasetArray, segmentId: Long)(implicit ec: ExecutionContext,
+                                                                                    tc: TokenContext): Fox[Long] =
+    for {
+      asMultiArray <- segmentToAgglomerate.readAsMultiArray(offset = segmentId, shape = 1)
+    } yield asMultiArray.getLong(0)
+
+  private def openZarrArrayCached(agglomerateFileKey: AgglomerateFileKey,
+                                  zarrArrayName: String)(implicit ec: ExecutionContext, tc: TokenContext) =
+    openArraysCache.getOrLoad((agglomerateFileKey, zarrArrayName),
+                              _ => openZarrArray(agglomerateFileKey, zarrArrayName))
+
+  private def openZarrArray(agglomerateFileKey: AgglomerateFileKey,
+                            zarrArrayName: String)(implicit ec: ExecutionContext, tc: TokenContext): Fox[DatasetArray] =
+    for {
+      groupVaultPath <- dataVaultService.getVaultPath(RemoteSourceDescriptor(agglomerateFileKey.attachment.path, None))
+      segmentToAgglomeratePath = groupVaultPath / zarrArrayName
+      zarrArray <- Zarr3Array.open(segmentToAgglomeratePath,
+                                   DataSourceId("dummy", "unused"),
```

Comment:
Afaik this is not the first place were we have a dummy datasource id. Maybe add a "dummy" def / val to the companion object? So it is clear when a dummy id is used and this also makes it easier to find the locations in the code where this is used :thinking:. I think this is lille cleaner this, but else has no real benefits.

Reply:
The other place is for volume tracings. There I use `DataSourceId("VolumeTracing", dataLayer.name)`, which I think is a fair difference. I’d keep this as it is for the moment.

Down the line, maybe we can introduce a layer of abstraction to open Zarr Arrays without supplying wk-specific stuff like AxisOrder and DataSourceId. But I don’t have a good plan of action for that yet.